keps/sig-api-machinery/NNNN-client-go-alternative-services/README.md

```diff
@@ -0,0 +1,739 @@
+
+All conformant clusters are required to publish a list of endpoints with the apiserver addresses.In the default implementation, this list of endpoints is generated by a reconcile loop that guarantees that only ready apiserver addresses are present, see https://github.com/kubernetes/kubernetes/tree/master/pkg/controlplane/reconcilers.
+
+The proposal is for API servers to have an option to enable the generation of the Alternative Services headers based on the list of Endpoints created for the `kubernetes.default` service, per example:
+
+```sh
+1103 12:30:59.066469 1558329 round_trippers.go:454] GET https://127.0.0.1:44267/api/v1/namespaces/default/pods?limit=500 200 OK in 1 milliseconds
+I1103 12:30:59.066484 1558329 round_trippers.go:460] Response Headers:
+I1103 12:30:59.066491 1558329 round_trippers.go:463]     Cache-Control: no-cache, private
+I1103 12:30:59.066502 1558329 round_trippers.go:463]     Alt-Svc: h2="10.0.0.2:6443", h2="10.0.0.3:6443", h2="10.0.0.4:6443
+```
+
+### Client side HA
+
+Client-go creates a layer on top of the golang net/http library abstracting the communications against the apiservers using [Requests](https://github.com/kubernetes/client-go/blob/master/rest/request.go).
+
+The golang net/http library allows to modify a request in different points:
+- RoundTripper/Transport
+- Dialer
+- DNS dialer
+
+The client-go base-URL is immutable after creation, that guarantees that Requests will not be able to modify it externally.
+
+Client-go already implements some [custom RoundTrippers](https://github.com/kubernetes/client-go/blob/master/transport/round_trippers.go) for some functionalities, like debugging, authentication, impersonation, ...
+
+The proposal is to implement a new round tripper that has a local cache with the list of available apiservers.
+This list can be created manually via configuration (comma-separated list on the Config object) or automatically via the Alt-Svc headers published.
+
+The round tripper behavior, if there are alternative services, will be:
+1. Prefer local (same system as the client) alternative services.
+2. Prefer the one in the original request if is listed as an alternative service.
+3. If one alternative service is chosen and succeed, stick to it, only change on network failures.
+4. If the connection against an alternative service fails, block that alternative service for a specified timeout so it will not be retried, retry against the other available alternative services, if everything fails fall back to the original host.
+
+It retries only in case of network failures, that guarantees the idempotency of the request.
+
+The round tripper modifies the original `URL.Host` field with the alternative service `Host` and sets the http2 pseudo header `:authority` field and the SNI TLS server name to `kubernetes.default`
```

Comment:
a client that can reach an internal or external load balancer may not have a CA to terminate this service name.

Reply:
this is a requirement L210
> To avoid possible security issues, this feature will require HTTP2 and HTTPS in order to work.

Reply:
```suggestion
The round tripper modifies the original `URL.Host` field with the alternative service `Host` and sets the http2 pseudo header `:authority` field and the SNI TLS server name to `kubernetes.default`
If the connection against the server fails because of a certificate error, that host is blocked forever and will not be retried anymore.
```